source/parser/binding.civet

```diff
@@ -140,9 +142,71 @@ function arrayElementHasTrailingComma(elementNode)
   const lastChild = elementNode.children.-1
   return lastChild and lastChild[lastChild.length - 1]?.token is ","
 
+// If an ObjectBindingPattern doesn't have a typeSuffix, this function
+// extracts one from typeSuffixes for the binding properties (if present).
+function gatherBindingPatternTypeSuffix(pattern: ArrayBindingPattern | ObjectBindingPattern): BindingPattern
+  return pattern if pattern.typeSuffix?
+  count .= 0
+  switch pattern.type
+    when "ArrayBindingPattern"
+      types: ASTNode[] :=
+        for each elem of pattern.elements
+          typeSuffix .= elem.typeSuffix
+          typeSuffix ??= elem.binding?.typeSuffix
+          count++ if typeSuffix
+          typeElement .= [ typeSuffix?.t, "," ]
```

Comment:
We should only add trailing commas as authored in the source. It's not necessary for this to merge but ideally trailing commas would always be as authored (some parts of the parser currently do this better than others).

Reply:
Thanks for encouraging this; it turned out to be pretty easy to improve upon using `delim` properties (just has to add them in one case). It's still not exactly perfect; there's an added comma at the end of an object when a `...rest` part got extracted out so it wouldn't be necessary. But it seems close enough.

Leaving better preservation of whitespace to future work. It's pretty good now, but definitely not perfect.